In [28]:
#Task 1
import pandas as pd
print(pd.__version__)

2.3.3


In [29]:
#Task 2
customers_df = pd.read_csv("customers.csv")
customers_df

,age,income,visits_per_month
0,30.483571,33893.885591,9.295377
1,35.615149,33126.773002,7.531726
2,35.896064,41139.477833,7.061051
3,30.712800,31292.658458,7.068540
4,29.209811,19693.758043,4.550164
...,...,...,...
295,67.926951,103758.922780,2.640708
296,59.134584,111630.553709,3.820482
297,58.550918,111000.120344,4.169296
298,64.675113,104730.648314,2.832882


In [30]:
# Task 3
print(customers_df.head(6))

         age        income  visits_per_month
0  30.483571  33893.885591          9.295377
1  35.615149  33126.773002          7.531726
2  35.896064  41139.477833          7.061051
3  30.712800  31292.658458          7.068540
4  29.209811  19693.758043          4.550164
5  25.188562  26897.351037          8.628495


In [31]:
print(customers_df.tail(4))

           age         income  visits_per_month
296  59.134584  111630.553709          3.820482
297  58.550918  111000.120344          4.169296
298  64.675113  104730.648314          2.832882
299  56.026996  113097.634498          2.183064


In [32]:
print(customers_df.sample(5))

           age         income  visits_per_month
228  46.445105  105694.976453          2.759675
267  51.387687   96441.811654          2.537725
238  54.736557  111549.528230          3.114228
210  43.152748  117943.254792          2.841992
85   25.578830   45135.289193          6.584661


Sample DOES return different rows each time because the results are random

In [33]:
#Task 4
cars_df = pd.read_csv("cars.csv")
cars_df

,mpg,cylinders,displacement,horsepower,weight,acceleration,model year,origin,car name
0,18.0,8,307.0,130,3504,12.0,70,1,chevrolet chevelle malibu
1,15.0,8,350.0,165,3693,11.5,70,1,buick skylark 320
2,18.0,8,318.0,150,3436,11.0,70,1,plymouth satellite
3,16.0,8,304.0,150,3433,12.0,70,1,amc rebel sst
4,17.0,8,302.0,140,3449,10.5,70,1,ford torino
...,...,...,...,...,...,...,...,...,...
393,27.0,4,140.0,86,2790,15.6,82,1,ford mustang gl
394,44.0,4,97.0,52,2130,24.6,82,2,vw pickup
395,32.0,4,135.0,84,2295,11.6,82,1,dodge rampage
396,28.0,4,120.0,79,2625,18.6,82,1,ford ranger


Rows?  -- 398
Columns? --9
Column Names? -- mpg	cylinders	displacement	horsepower	weight	acceleration	model year	origin	car name

In [34]:
# Task 5
cars_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 398 entries, 0 to 397
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   mpg           398 non-null    float64
 1   cylinders     398 non-null    int64  
 2   displacement  398 non-null    float64
 3   horsepower    398 non-null    object 
 4   weight        398 non-null    int64  
 5   acceleration  398 non-null    float64
 6   model year    398 non-null    int64  
 7   origin        398 non-null    int64  
 8   car name      398 non-null    object 
dtypes: float64(3), int64(4), object(2)
memory usage: 28.1+ KB


Task 5
Whole Numbers: cylinders, weight, model year, origin
Decimal Numbers: mpg, displacement, acceleration
Object: horsepower, car name

In [35]:
#Task 6
titanic_df = pd.read_csv("titanic_survival.csv")
titanic_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 15 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   survived     891 non-null    int64  
 1   pclass       891 non-null    int64  
 2   sex          891 non-null    object 
 3   age          714 non-null    float64
 4   sibsp        891 non-null    int64  
 5   parch        891 non-null    int64  
 6   fare         891 non-null    float64
 7   embarked     889 non-null    object 
 8   class        891 non-null    object 
 9   who          891 non-null    object 
 10  adult_male   891 non-null    bool   
 11  deck         203 non-null    object 
 12  embark_town  889 non-null    object 
 13  alive        891 non-null    object 
 14  alone        891 non-null    bool   
dtypes: bool(2), float64(2), int64(4), object(7)
memory usage: 92.4+ KB


In [36]:
Rows = 891  columns = 15
most missing values: deck
You know it is missing because of the number of non-null values in the deck column is 203, which is lower than the number of rows...by a lot

SyntaxError: invalid syntax (170719289.py, line 1)

In [ ]:
#Task 7
cars_df.describe()

,mpg,cylinders,displacement,weight,acceleration,model year,origin
count,398.000000,398.000000,398.000000,398.000000,398.000000,398.000000,398.000000
mean,23.514573,5.454774,193.425879,2970.424623,15.568090,76.010050,1.572864
std,7.815984,1.701004,104.269838,846.841774,2.757689,3.697627,0.802055
min,9.000000,3.000000,68.000000,1613.000000,8.000000,70.000000,1.000000
25%,17.500000,4.000000,104.250000,2223.750000,13.825000,73.000000,1.000000
50%,23.000000,4.000000,148.500000,2803.500000,15.500000,76.000000,1.000000
75%,29.000000,8.000000,262.000000,3608.000000,17.175000,79.000000,2.000000
max,46.600000,8.000000,455.000000,5140.000000,24.800000,82.000000,3.000000


Mean of mpg = 23.514573
Max horsepower = ??
Median weight of car: 2803.5

In [ ]:
#Task 8
titanic_df['fare'].describe()

count    891.000000
mean      32.204208
std       49.693429
min        0.000000
25%        7.910400
50%       14.454200
75%       31.000000
max      512.329200
Name: fare, dtype: float64

The fare data has outliers. You can tell because the mean and median are pretty different.

In [ ]:
#Task 9
import pandas as pd
data = {
    'product': ['pencil', 'pen', 'eraser', 'notebook'],
    'price': [1, 5, 2, 9],
    'in_stock': [True, True, False, True]
}

data_df = pd.DataFrame(data)
data_df.dtypes


product     object
price        int64
in_stock      bool
dtype: object

#Task 10
1. Shape tells you how many rows and columns there are. Info provides additional details like non-null counts, column names, dtype, and index. You would use shape to get a surface level view of the data (make sure file is read in correctly or modified)
2. head provides the specified number of rows from the beginning of the data set. Sample provides a random sample from the data. Sample is better when your data is grouped by row. 
3. Deck would be missing if this was a less documented item.


In [38]:
#Task 11
print(titanic_df['age'].describe())

count    714.000000
mean      29.699118
std       14.526497
min        0.420000
25%       20.125000
50%       28.000000
75%       38.000000
max       80.000000
Name: age, dtype: float64


Task 11 markdown
1. Youngest age = .42 years
2. oldest age = 80
3. count row tells you how many rows have data populated in that column. It is not the same, because the data is incomplete


In [39]:
# Task 12
cars_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 398 entries, 0 to 397
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   mpg           398 non-null    float64
 1   cylinders     398 non-null    int64  
 2   displacement  398 non-null    float64
 3   horsepower    398 non-null    object 
 4   weight        398 non-null    int64  
 5   acceleration  398 non-null    float64
 6   model year    398 non-null    int64  
 7   origin        398 non-null    int64  
 8   car name      398 non-null    object 
dtypes: float64(3), int64(4), object(2)
memory usage: 28.1+ KB


In [40]:
customers_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 300 entries, 0 to 299
Data columns (total 3 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   age               300 non-null    float64
 1   income            300 non-null    float64
 2   visits_per_month  300 non-null    float64
dtypes: float64(3)
memory usage: 7.2 KB


1. Cars is the larger of the data sets at 398 rows 
2. Cars has more columns 9
3. Cars has incorrect values : horsepower should be number and not object

In [45]:
#Task 13
cars_df.sample(5, random_state=42)

,mpg,cylinders,displacement,horsepower,weight,acceleration,model year,origin,car name
198,33.0,4,91.0,53,1795,17.4,76,3,honda civic
396,28.0,4,120.0,79,2625,18.6,82,1,ford ranger
33,19.0,6,232.0,100,2634,13.0,71,1,amc gremlin
208,13.0,8,318.0,150,3940,13.2,76,1,plymouth volare premier v8
93,14.0,8,318.0,150,4237,14.5,73,1,plymouth fury gran sedan


In [49]:
cars_df.sample(5)

,mpg,cylinders,displacement,horsepower,weight,acceleration,model year,origin,car name
224,15.0,8,302.0,130,4295,14.9,77,1,mercury cougar brougham
59,23.0,4,97.0,54,2254,23.5,72,2,volkswagen type 3
342,30.0,4,135.0,84,2385,12.9,81,1,plymouth reliant
12,15.0,8,400.0,150,3761,9.5,70,1,chevrolet monte carlo
161,16.0,6,250.0,105,3897,18.5,75,1,chevroelt chevelle malibu


Task 13 questions
1. without the random_state assigned, the rows change each run
2. You can compare the same random rows when making changes to the data


In [55]:
# Task 14
def load_and_check(file_path):
    df = pd.read_csv(file_path)
    rows, columns = df.shape
    print(f'Loaded: {rows} rows, {columns} columns')
    return df

load_and_check("titanic_survival.csv")
load_and_check("cars.csv")
load_and_check("customers.csv")


Loaded: 891 rows, 15 columns
Loaded: 398 rows, 9 columns
Loaded: 300 rows, 3 columns


,age,income,visits_per_month
0,30.483571,33893.885591,9.295377
1,35.615149,33126.773002,7.531726
2,35.896064,41139.477833,7.061051
3,30.712800,31292.658458,7.068540
4,29.209811,19693.758043,4.550164
...,...,...,...
295,67.926951,103758.922780,2.640708
296,59.134584,111630.553709,3.820482
297,58.550918,111000.120344,4.169296
298,64.675113,104730.648314,2.832882


In [61]:
#Task 15
cars_df.sample(20)

,mpg,cylinders,displacement,horsepower,weight,acceleration,model year,origin,car name
215,13.0,8,318.0,150,3755,14.0,76,1,dodge d100
382,34.0,4,108.0,70,2245,16.9,82,3,toyota corolla
226,20.5,6,231.0,105,3425,16.9,77,1,buick skylark
128,15.0,6,250.0,100,3336,17.0,74,1,chevrolet nova
388,26.0,4,156.0,92,2585,14.5,82,1,chrysler lebaron medallion
234,24.5,4,151.0,88,2740,16.0,77,1,pontiac sunbird coupe
297,25.4,5,183.0,77,3530,20.1,79,2,mercedes benz 300d
337,32.4,4,107.0,72,2290,17.0,80,3,honda accord
37,18.0,6,232.0,100,3288,15.5,71,1,amc matador
116,16.0,8,400.0,230,4278,9.5,73,1,pontiac grand prix


Task 15
There are question marks in the horsepower column causing the object type.

In [66]:
#Task 16
titanic_df = pd.read_csv("titanic_survival.csv")
titanic_df.shape



(891, 15)

In [65]:
titanic_df.head(5)

,survived,pclass,sex,age,sibsp,parch,fare,embarked,class,who,adult_male,deck,embark_town,alive,alone
0,0,3,male,22.0,1,0,7.2500,S,Third,man,True,NaN,Southampton,no,False
1,1,1,female,38.0,1,0,71.2833,C,First,woman,False,C,Cherbourg,yes,False
2,1,3,female,26.0,0,0,7.9250,S,Third,woman,False,NaN,Southampton,yes,True
3,1,1,female,35.0,1,0,53.1000,S,First,woman,False,C,Southampton,yes,False
4,0,3,male,35.0,0,0,8.0500,S,Third,man,True,NaN,Southampton,no,True


In [68]:
titanic_df.dtypes

survived         int64
pclass           int64
sex             object
age            float64
sibsp            int64
parch            int64
fare           float64
embarked        object
class           object
who             object
adult_male        bool
deck            object
embark_town     object
alive           object
alone             bool
dtype: object

In [69]:
titanic_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 15 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   survived     891 non-null    int64  
 1   pclass       891 non-null    int64  
 2   sex          891 non-null    object 
 3   age          714 non-null    float64
 4   sibsp        891 non-null    int64  
 5   parch        891 non-null    int64  
 6   fare         891 non-null    float64
 7   embarked     889 non-null    object 
 8   class        891 non-null    object 
 9   who          891 non-null    object 
 10  adult_male   891 non-null    bool   
 11  deck         203 non-null    object 
 12  embark_town  889 non-null    object 
 13  alive        891 non-null    object 
 14  alone        891 non-null    bool   
dtypes: bool(2), float64(2), int64(4), object(7)
memory usage: 92.4+ KB


In [70]:
titanic_df.describe()

,survived,pclass,age,sibsp,parch,fare
count,891.000000,891.000000,714.000000,891.000000,891.000000,891.000000
mean,0.383838,2.308642,29.699118,0.523008,0.381594,32.204208
std,0.486592,0.836071,14.526497,1.102743,0.806057,49.693429
min,0.000000,1.000000,0.420000,0.000000,0.000000,0.000000
25%,0.000000,2.000000,20.125000,0.000000,0.000000,7.910400
50%,0.000000,3.000000,28.000000,0.000000,0.000000,14.454200
75%,1.000000,3.000000,38.000000,1.000000,0.000000,31.000000
max,1.000000,3.000000,80.000000,8.000000,6.000000,512.329200


Task 16 
The titanic_survival dataset contains information on 891 passengers that traveled on the titanic and what the outcome of that journey was. The data set contains the columns survived, passenger class, sex, age etc. Some information was not documented and not all of the columns are filled out. The data can inform you of which factors impacted survival rates.